File: 2015/go/day01/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	currentFloor := part1_GetCurrentFloor()
	fmt.Println(currentFloor)

	position := part2_GetPosition()
	fmt.Println(position)
}

func part1_GetCurrentFloor() int {
	input, _ := os.Open("../../inputs/day01.txt")
	defer input.Close()

	s := bufio.NewScanner(input)
	s.Split(bufio.ScanRunes)

	currentFloor := 0
	for s.Scan() {
		if s.Text() == "(" {
			currentFloor += 1
		} else if s.Text() == ")" {
			currentFloor -= 1
		}
	}

	return currentFloor
}

func part2_GetPosition() int {
	input, _ := os.Open("../../inputs/day01.txt")
	defer input.Close()

	s := bufio.NewScanner(input)
	s.Split(bufio.ScanRunes)

	position := 1
	currentFloor := 0
	for s.Scan() {
		if s.Text() == "(" {
			currentFloor += 1
		} else if s.Text() == ")" {
			currentFloor -= 1
		}

		if currentFloor == -1 {
			break
		}
		position += 1
	}

	return position
}

File: 2015/rs/day01/src/main.rs
use std::io::{BufRead, BufReader};
use std::fs::File;

fn main() {
    let current_floor = part1_get_current_floor();
    let position = part2_get_position();
    println!("current floor: {}, position: {}", current_floor, position);
}

fn part1_get_current_floor() -> i32 {
    let file = File::open("../../inputs/day01.txt").expect("unable to open input file");
    let reader = BufReader::new(file);

    let mut current_floor = 0;
    for line in reader.lines() {
        for ch in line.expect("unable to read line").chars() {
            if ch == '(' {
                current_floor += 1;
            } else if ch == ')' {
                current_floor -= 1;
            }
        }
    }

    current_floor
}

fn part2_get_position() -> i32 {
    let file = File::open("../../inputs/day01.txt").expect("unable to open input file");
    let reader = BufReader::new(file);

    let mut current_floor = 0;
    let mut position = 1;
    for line in reader.lines() {
        for ch in line.expect("unable to read line").chars() {
            if ch == '(' {
                current_floor += 1;
            } else if ch == ')' {
                current_floor -= 1;
            }

            if current_floor == -1 {
                break;
            }
            position += 1;
        }
    }

    position
}